src/lua/version.lua

```diff
@@ -0,0 +1,243 @@
+--
+-- Version parser adopted to the Tarantool's versioning policy:
+-- https://www.tarantool.io/en/doc/latest/release/policy/#versioning-policy
+--
+-- Tarantool's versions use SemVer-like format:
+--     x.y.z-typeN-commit-ghash
+--
+-- * x.y.z - major, minor, patch release numbers;
+-- * typeN - pre-release type and its optional number: alpha1, rc10. Optional;
+-- * commit - number of commits since the latest release. Optional;
+-- * ghash - latest commit hash in format g<hash>. Optional.
+--
+-- Differences with the semver docs:
+--
+-- * Pre-release identifiers are not separated. Not 'alpha.1' but 'alpha1'.
+-- * Pre-release identifiers cannot be nested. Not x.y.z-alpha.beta.
+-- * Pre-release identifier always includes non-numeric type: e.g. beta, rc.
+-- * Build metadata is denoted by appending hyphen, not plus sign. Not
+--   x.y.z-typeN+commit-ghash, but x.y.z-typeN-commit-ghash.
+--
+-- Note, that ghash itself and everything after it (e.g. -dev suffix) is
+-- ignored and doesn't affect comparing of the versions.
+--
+
+local ffi   = require('ffi')
+local utils = require('internal.utils')
+
+local check_param = utils.check_param
+
+--
+-- Cdata is used in order to allow comparison of the version object with
+-- strings. Since relational metamethods do not support mixed types in Lua,
+-- using tables is not possible.
+--
+ffi.cdef([[
+    struct version {
+        /** Major number. */
+        uint32_t major;
+        /** Minor number. */
+        uint32_t minor;
+        /** Patch number. */
+        uint32_t patch;
+        /** Pre-release type: e.g. entrypoint, apha. */
+        const char *_release_type_pointer;
+        /** Pre-release number: e.g. alpha1. */
+        uint16_t release_num;
+        /** Number of commits since release. */
+        uint32_t commit_num;
+    };
+]])
+
+local release_type_weight = {
+    entrypoint = 10,
+    alpha = 20,
+    beta = 30,
+    rc = 40,
+}
+
+local function version_new(major, minor, patch, rel_type, rel_num, commit)
+    if rel_type and not release_type_weight[rel_type] then
+        local msg = "Unknown release type '%s'"
+        box.error(box.error.ILLEGAL_PARAMS, string.format(msg, rel_type))
+    end
+
+    rel_type = rel_type or ''
+    rel_num = rel_num or 0
+    commit = commit or 0
+    check_param(major, 'major', 'number')
+    check_param(minor, 'minor', 'number')
+    check_param(patch, 'patch', 'number')
+    check_param(rel_type, 'release type', 'string')
+    check_param(rel_num, 'release number', 'number')
```

Comment:
It is about a pre-release and have nothing related to a release. I find these names confusing.

Also, I don't like that these fields have two-word names unlike all the others.

semver.org calls this part `<pre-release identifier>` in a grammar and just `prerelease` in a regexp. Even just `suffix` would make less confusion, IMHO.

Are we really need to split `rc1` to `rc` and `1` in some use scenario? If we don't, I would union them and call `prerelease`. This is simple and straightforward.

Reply:
I think we could rename `commit_num` to `commit`. As for `release_num`, I don't like it, either. I agree that having just one field `prerelease` (alpha, rc1, rc2) would look better. We just need to be careful when comparing these identifiers.

Reply:
Now there's only `prerelease` field, which returns the concatenated `prerelease_name` and `prerelease_num`. User still can get separated values via `_prerelease_name` and `_prerelease_num` if he'll look into sources